File: prog/c++/copy_issues.cpp
#include <iostream>
#include <complex>
#include <vector>
#include <map>
#include <set>

void demonstrateIssues() {
    // Shallow Copy
    int *a = new int(10);
    int *b = a;   // shallow copy: both point to same memory

    *b = 20;      // changing b also changes a
    std::cout << "Shallow Copy:" << "\n";
    std::cout << "a = " << *a << ", b = " << *b << "\n";  // both 20

    // Deep Copy
    int *x = new int(10);
    int *y = new int(*x);  // deep copy: new memory with same value

    *y = 30;      // changing y does not affect x
    std::cout << "\nDeep Copy:" << "\n";
    std::cout << "x = " << *x << ", y = " << *y << "\n";  // x=10, y=30

    delete a; // free memory
    // delete b;  // dangerous in shallow copy (same memory as a)
    delete x;
    delete y;


    // Immutable types
    int xx = 42;
    int yy = xx;  // Copy by value
    yy += 1;
    std::cout << "Integer: " << xx << " " << yy << "\n";

    double z = 3.14;
    double w = z;  // Copy by value
    w += 1.0;
    std::cout << "Float: " << z << " " << w << "\n";

    std::complex<double> c(2, 3);
    std::complex<double> d = c;  // Copy by value
    d += std::complex<double>(1, 0);
    std::cout << "Complex: " << c << " " << d << "\n";

    // Mutable types
    std::vector<int> lst = {1, 2, 3};
    std::vector<int> lst_copy = lst;  // Deep copy by default
    lst_copy[0] = 2;
    std::cout << "List: " << lst[0] << " " << lst_copy[0] << "\n";

    std::map<std::string, std::string> dct = {{"key", "value"}};
    std::map<std::string, std::string> dct_copy = dct;  // Deep copy
    dct_copy["key"] = "new_value";
    std::cout << "Dict: " << dct["key"] << " " << dct_copy["key"] << "\n";

    std::set<int> st = {1, 2, 3};
    std::set<int> st_copy = st;  // Deep copy
    st_copy.insert(42);
    std::cout << "Set: " << *st.begin() << " " << *st_copy.begin() << "\n";
    for (auto& e : st)
        std::cout << e;
    std::cout << "\n";
    for (auto& e : st_copy)
        std::cout << e;
}

int main() {
    demonstrateIssues();
    return 0;
}

File: prog/python/copy_issues.py
from copy import deepcopy

def demonstrate_issues():
    # Immutable types
    x = 42
    y = x  # Copy by reference, but integers are immutable
    y += 1
    print("Integer:", x, y)  # x is unaffected

    z = 3.14
    w = z  # Same for floats
    w += 1.0
    print("Float:", z, w)

    c = 2 + 3j
    d = c  # Complex numbers are immutable
    d += 1
    print("Complex:", c, d)

    # Mutable types
    lst = [1, 2, 3]
    lst_copy = lst  # Shallow copy, shared reference
    lst_copy[0] = 42
    print("List:", lst, lst_copy)  # Both are modified

    dct = {"key": "value"}
    dct_copy = dct  # Shallow copy, shared reference
    dct_copy["key"] = "new_value"
    print("Dict:", dct, dct_copy)

    st = {1, 2, 3}
    st_copy = st  # Shallow copy, shared reference
    st_copy.add(42)
    print("Set:", st, st_copy)


demonstrate_issues()

def modify_list(input_list):
    input_list[0] = 42

original_list = [1, 2, 3]

# Pass a copy instead
modify_list(deepcopy(original_list))

print("Original List:", original_list)  # Original list remains unchanged
